ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/FailedExecutionPool.java

```diff
@@ -0,0 +1,121 @@
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.exceptions.ExceptionUtil;
+import tech.pegasys.teku.infrastructure.logging.LogFormatter;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.logic.common.statetransition.results.BlockImportResult;
+import tech.pegasys.teku.storage.server.ShuttingDownException;
+
+public class FailedExecutionPool {
+  private static final Logger LOG = LogManager.getLogger();
+  static final Duration MAX_RETRY_DELAY = Duration.ofSeconds(30);
+  static final Duration SHORT_DELAY = Duration.ofSeconds(2);
+  private final Queue<SignedBeaconBlock> awaitingExecution = new ArrayBlockingQueue<>(10);
+  private final BlockManager blockManager;
+  private final AsyncRunner asyncRunner;
+
+  private Optional<SignedBeaconBlock> retryingBlock = Optional.empty();
+
+  private Duration currentDelay = SHORT_DELAY;
+
+  public FailedExecutionPool(final BlockManager blockManager, final AsyncRunner asyncRunner) {
+    this.blockManager = blockManager;
+    this.asyncRunner = asyncRunner;
+  }
+
+  public synchronized void addFailedBlock(final SignedBeaconBlock block) {
+    if (retryingBlock.isEmpty()) {
+      retryingBlock = Optional.of(block);
+      scheduleNextRetry(block);
+    } else {
+      if (retryingBlock.get().equals(block) || awaitingExecution.contains(block)) {
+        // Already retrying this block.
+        return;
+      }
+      if (!awaitingExecution.offer(block)) {
+        LOG.info(
+            "Discarding block {} as execution retry pool capacity exceeded",
+            LogFormatter.formatBlock(block.getSlot(), block.getRoot()));
+      }
+    }
+  }
+
+  private synchronized void handleExecutionResult(
+      final SignedBeaconBlock block, final BlockImportResult importResult) {
+    if (importResult.hasFailedExecutingExecutionPayload()) {
+      currentDelay = currentDelay.multipliedBy(2);
+      if (currentDelay.compareTo(MAX_RETRY_DELAY) > 0) {
+        currentDelay = MAX_RETRY_DELAY;
+      }
+      if (awaitingExecution.isEmpty() || isTimeout(importResult)) {
+        scheduleNextRetry(block);
+      } else {
+        // Try a different block
+        final SignedBeaconBlock nextBlock = awaitingExecution.remove();
+        awaitingExecution.add(block);
+        retryingBlock = Optional.of(nextBlock);
+        scheduleNextRetry(nextBlock);
+      }
+    } else {
+      currentDelay = SHORT_DELAY;
+      retryingBlock = Optional.ofNullable(awaitingExecution.poll());
+      retryingBlock.ifPresent(this::retryExecution);
+    }
+  }
+
+  private boolean isTimeout(final BlockImportResult importResult) {
+    return importResult
+        .getFailureCause()
+        .map(
+            error ->
+                ExceptionUtil.hasCause(error, TimeoutException.class)
+                    || ExceptionUtil.hasCause(error, SocketTimeoutException.class))
```

Comment:
If I correctly understand it could be also websockets and ipc there. For Websockets it looks like timeout will be `IOException`. Anyway, this method is detached and could be improved later

Reply:
hmm, interesting. The exception I got from OkHttp was an `InterruptedIOException` with a `SocketTimeoutException` as the cause.  Probably reasonable to look for InterruptedIOException directly though.